providers/discord.go

```diff
@@ -0,0 +1,147 @@
+package providers
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/requests"
+)
+
+// DiscordProvider represents an Discord based Identity Provider
+type DiscordProvider struct {
+	*ProviderData
+}
+
+var _ Provider = (*DiscordProvider)(nil)
+
+const (
+	discordProviderName = "Discord"
+	discordDefaultScope = "identify email guilds"
+)
+
+var (
+	// Default Login URL for Discord.
+	// Pre-parsed URL of https://discord.com/api/oauth2/authorize.
+	discordDefaultLoginURL = &url.URL{
+		Scheme: "https",
+		Host:   "discord.com",
+		Path:   "/api/oauth2/authorize",
+	}
+
+	// Default Redeem URL for Discord.
+	// Pre-parsed URL of https://discord.com/api/oauth2/token
+	discordDefaultRedeemURL = &url.URL{
+		Scheme: "https",
+		Host:   "discord.com",
+		Path:   "/api/oauth2/token",
+	}
+
+	// Default Validate URL for Discord.
+	// Pre-parsed URL of https://discord.com/api/users/@me
+	discordDefaultValidateURL = &url.URL{
+		Scheme: "https",
+		Host:   "discord.com",
+		Path:   "/api/users/@me",
+	}
+
+	// Default Validate URL for Discord.
+	// Pre-parsed URL of https://discord.com/api/users/@me/guilds
+	discordGuildURL = &url.URL{
+		Scheme: "https",
+		Host:   "discord.com",
+		Path:   "/api/users/@me/guilds",
+	}
+)
+
+// NewDiscordProvider initiates a new DiscordProvider
+func NewDiscordProvider(p *ProviderData, opts options.DiscordOptions) *DiscordProvider {
+	p.setProviderDefaults(providerDefaults{
+		name:        discordProviderName,
+		loginURL:    discordDefaultLoginURL,
+		redeemURL:   discordDefaultRedeemURL,
+		profileURL:  discordDefaultValidateURL,
+		validateURL: discordDefaultValidateURL,
+		scope:       discordDefaultScope,
+	})
+	p.setAllowedGroups(opts.Guilds)
```

Comment:
This line is a problem with the legacy `--allowed-group` flag or `allowed_groups` config field as it will always overwrite the set values with an empty map `map[string]struct {} []`

Reply:
unfortunately, you will have to add legacy support

Reply:
Can confirm that legacy support is broken right now. --allowed-group doesn't work